controllers/backup_test.go

```diff
@@ -0,0 +1,259 @@
+				Name:      "test",
+				Namespace: "f1",
+				Labels:    newSource("f1", "test", "1G").Labels,
+				Spec:      newSource("f1", "test", "1G").Spec,
+			},
+			in: state{
+				source: nil,
+			},
+			out:        state{},
+			targetSize: "4G",
+			err:        errCritical,
+		},
+		"abort, if too small backup exists": {
+			namespace: "f2",
+			syncImage: "blub",
+			in: state{
+				source: newSource("f2", "test", "2G"),
+				backup: newBackup("f2", "test-backup-2g", "1G"),
+			},
+			out:        state{},
+			targetSize: "4G",
+			err:        errAbort,
+		},
+		"abort, if job failed": {
+			namespace: "f3",
+			syncImage: "blub",
+			in: state{
+				source: newSource("f3", "test", "1G"),
+				backup: newBackup("f3", "test-backup-2g", "1G"),
+				job: newTestJob("f3",
+					client.ObjectKey{Namespace: "f3", Name: "test"},
+					client.ObjectKey{Namespace: "f3", Name: "test-backup-1g"},
+					"blub", &jobFailed),
+			},
+			out:        state{},
+			targetSize: "4G",
+			err:        errAbort,
+		},
+	}
+
+	for k, tc := range tcs {
+		tc := tc
+		ctx := context.Background()
+		k := k
+
+		t.Run(k, func(t *testing.T) {
+			t.Parallel()
+			assert := assert.New(t)
+			require := require.New(t)
+			require.Nil(k8sClient.Create(ctx, &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{Name: tc.namespace},
+			}))
+			if tc.in.source != nil {
+				require.Nil(k8sClient.Create(ctx, tc.in.source))
+			}
+			if tc.in.backup != nil {
+				require.Nil(k8sClient.Create(ctx, tc.in.backup))
+			}
+			if tc.in.job != nil {
+				stat := tc.in.job.Status
+				require.Nil(k8sClient.Create(ctx, tc.in.job))
+				tc.in.job.Status = stat // Create removes status
+				require.Nil(k8sClient.Status().Update(ctx, tc.in.job))
+			}
+			r := StatefulSetReconciler{
+				Client:             k8sClient,
+				Scheme:             k8sClient.Scheme(),
+				SyncContainerImage: tc.syncImage,
+			}
+			pi := pvcInfo{}
+			if tc.in.source != nil {
+				pi = pvcInfo{
+					Name:      tc.in.source.Name,
+					Namespace: tc.in.source.Namespace,
+					Labels:    tc.in.source.Labels,
+					Spec:      tc.in.source.Spec,
+				}
+			}
+			if tc.pvcInfo != nil {
+				pi = *tc.pvcInfo
+			}
+			pi.TargetSize = resource.MustParse(tc.targetSize)
+
+			err := r.backupPVC(ctx, pi)
+
+			if tc.err != nil {
+				require.ErrorIs(err, tc.err)
+				return
+			}
+			if !tc.done {
+				require.ErrorIs(err, errInProgress)
+			} else {
+				require.Nil(err)
+			}
+
+			if tc.out.source != nil {
+				assert.Eventually(func() bool {
+					return pvcExists(ctx, k8sClient, tc.out.source)
+				}, timeout, interval, "Source is not as expected")
+			} else if tc.in.source != nil {
+				assert.Eventually(func() bool {
+					return pvcNotExists(ctx, k8sClient, tc.in.source)
+				}, timeout, interval, "Source not deleted")
+			}
+			if tc.out.backup != nil {
+				assert.Eventually(func() bool {
+					return pvcExists(ctx, k8sClient, tc.out.backup)
+				}, timeout, interval, "Backup is not as expected")
```

Comment:
Are you sure that much logic is needed in a single test? I think it makes sense to split some that are more grouped together instead of a one-size-fits-all test case?